File: packages/domain/src/measurePresets.ts
/**
 * Commercial measure presets on Module (H09 / #104).
 * Structure is parametric; the furniture owns the sellable size list.
 */

import { ValidationError } from './errors';
import type { DimensionPreset, FurnitureType, Module, Project } from './types';

export function moduleHasMeasurePresets(module: Module): boolean {
  return (module.presets?.length ?? 0) > 0;
}

/**
 * Resolve the commercial measure preset for a quote line.
 * When the module has presets, measurePresetId is required and must match.
 * When no presets, returns undefined (fixed / default dims path).
 */
export function resolveModuleMeasurePreset(
  module: Module,
  measurePresetId: string | undefined,
): DimensionPreset | undefined {
  const presets = module.presets ?? [];
  if (presets.length === 0) {
    if (measurePresetId?.trim()) {
      throw new ValidationError(
        `El mueble "${module.name}" (${module.code}) no define presets de medida; no se puede elegir una medida.`,
        {
          moduleCode: module.code,
          measurePresetId,
          field: 'measurePresetId',
        },
      );
    }
    return undefined;
  }

  if (!measurePresetId?.trim()) {
    throw new ValidationError(
      `Elegí un preset de medida para el mueble "${module.name}" (${module.code}).`,
      {
        moduleCode: module.code,
        field: 'measurePresetId',
      },
    );
  }

  const matched = presets.find((p) => p.id === measurePresetId);
  if (!matched) {
    throw new ValidationError(
      `El preset de medida no es válido para el mueble "${module.name}" (${module.code}).`,
      {
        moduleCode: module.code,
        measurePresetId,
        field: 'measurePresetId',
      },
    );
  }

  if (matched.width <= 0 || matched.height <= 0 || matched.depth <= 0) {
    throw new ValidationError(
      'Las dimensiones del preset deben ser mayores a 0',
      {
        moduleCode: module.code,
        measurePresetId: matched.id,
        field: 'presets',
      },
    );
  }

  return matched;
}

/** Default preset id when adding a line (first commercial option). */
export function defaultMeasurePresetId(module: Module): string | undefined {
  return module.presets?.[0]?.id;
}

/**
 * Pick the module preset whose depth/height best match the project defaults
 * for the module's furniture type (#109 / H14).
 *
 * - Resolves the module type (`module.furnitureType ?? 'inferior'`).
 * - Reads the matching entry from `defaults`. If none, or no depth/height set,
 *   falls back to the first preset (== `defaultMeasurePresetId`).
 * - Otherwise picks the preset with the smallest absolute-mm distance, summed
 *   over the dimensions present in the project default (depth and/or height).
 *   Ties keep the first preset in definition order (stable).
 * - Returns `undefined` when the module has no presets (fixed-dims path).
 */
export function pickPresetByMeasureDefaults(
  module: Module,
  defaults?: Project['measureDefaults'] | null,
): string | undefined {
  const presets = module.presets ?? [];
  const first = presets[0];
  if (!first) return undefined;

  const type: FurnitureType = module.furnitureType ?? 'inferior';
  const target = defaults?.[type];
  const wantDepth = typeof target?.depth === 'number';
  const wantHeight = typeof target?.height === 'number';
  if (!wantDepth && !wantHeight) {
    return first.id;
  }

  let best = first;
  let bestDist = distanceFor(best, target, wantDepth, wantHeight);
  for (let i = 1; i < presets.length; i++) {
    const candidate = presets[i];
    if (!candidate) continue;
    const d = distanceFor(candidate, target, wantDepth, wantHeight);
    if (d < bestDist) {
      best = candidate;
      bestDist = d;
    }
  }
  return best.id;
}

function distanceFor(
  preset: DimensionPreset,
  target: { readonly depth?: number; readonly height?: number },
  wantDepth: boolean,
  wantHeight: boolean,
): number {
  let dist = 0;
  if (wantDepth && typeof target.depth === 'number') {
    dist += Math.abs(preset.depth - target.depth);
  }
  if (wantHeight && typeof target.height === 'number') {
    dist += Math.abs(preset.height - target.height);
  }
  return dist;
}

export function validateModulePresets(module: Module): void {
  if (!module.presets) return;
  for (const preset of module.presets) {
    if (!preset.id?.trim()) {
      throw new ValidationError('Cada preset de medida necesita un id', {
        moduleCode: module.code,
        field: 'presets',
      });
    }
    if (preset.width <= 0 || preset.height <= 0 || preset.depth <= 0) {
      throw new ValidationError(
        'Las dimensiones del preset deben ser mayores a 0',
        {
          moduleCode: module.code,
          presetId: preset.id,
          field: 'presets',
        },
      );
    }
  }
  const ids = new Set<string>();
  for (const preset of module.presets) {
    if (ids.has(preset.id)) {
      throw new ValidationError(
        `Preset de medida duplicado: ${preset.id}`,
        {
          moduleCode: module.code,
          presetId: preset.id,
          field: 'presets',
        },
      );
    }
    ids.add(preset.id);
  }
}

File: backend-go/internal/domain/engine/measure_presets.go
package engine

import (
	"encoding/json"

	"github.com/tiagofur/muebles-backend/internal/domain"
)

// MeasureDefaultDims is one entry of Project.MeasureDefaults (#109).
// Pointers distinguish unset (nil) from an explicit zero.
type MeasureDefaultDims struct {
	Depth  *float64 `json:"depth,omitempty"`
	Height *float64 `json:"height,omitempty"`
}

// PickPresetByMeasureDefaults picks the module preset whose depth/height best
// match the project defaults for the module's furniture type (#109 / H14).
//
//   - Resolves the module type (Module.FurnitureType, "" → "inferior").
//   - Reads the matching entry from defaults. If none, or no depth/height set,
//     returns the first preset (== default preset).
//   - Otherwise picks the preset with the smallest absolute-mm distance summed
//     over the dimensions present in the project default.
//   - Returns "" when the module has no presets.
//
// defaultsRaw is the raw JSON of Project.MeasureDefaults
// ({"inferior":{"depth":560,"height":720}, ...}). nil/empty → no defaults.
func PickPresetByMeasureDefaults(mod domain.Module, defaultsRaw json.RawMessage) string {
	presets := mod.Presets
	if len(presets) == 0 {
		return ""
	}
	ft := mod.FurnitureType
	if ft == "" {
		ft = "inferior"
	}
	target, ok := parseMeasureDefaultForType(defaultsRaw, ft)
	if !ok || (target.Depth == nil && target.Height == nil) {
		return presets[0].ID
	}

	bestIdx := 0
	bestDist := distanceFor(presets[0], target)
	for i := 1; i < len(presets); i++ {
		d := distanceFor(presets[i], target)
		if d < bestDist {
			bestIdx = i
			bestDist = d
		}
	}
	return presets[bestIdx].ID
}

func distanceFor(preset domain.DimensionPreset, target MeasureDefaultDims) float64 {
	var dist float64
	if target.Depth != nil {
		dist += absFloat(float64(preset.DepthMm) - *target.Depth)
	}
	if target.Height != nil {
		dist += absFloat(float64(preset.HeightMm) - *target.Height)
	}
	return dist
}

func absFloat(v float64) float64 {
	if v < 0 {
		return -v
	}
	return v
}

// parseMeasureDefaultForType extracts the dims for a single furnitureType from
// the raw measure_defaults JSON. Returns ok=false when absent or invalid.
func parseMeasureDefaultForType(raw json.RawMessage, ft string) (MeasureDefaultDims, bool) {
	if len(raw) == 0 {
		return MeasureDefaultDims{}, false
	}
	var all map[string]MeasureDefaultDims
	if err := json.Unmarshal(raw, &all); err != nil {
		return MeasureDefaultDims{}, false
	}
	entry, ok := all[ft]
	if !ok {
		return MeasureDefaultDims{}, false
	}
	return entry, true
}
